numcodecs/compat.py

```diff
@@ -27,80 +25,150 @@
-    else:
-        return memoryview(v).tobytes()
-
-
-def buffer_copy(buf, out=None):
+def memory_copy(buf, out=None):
     """Copy the contents of the memory buffer from `buf` to `out`."""
 
     if out is None:
         # no-op
         return buf
 
-    # handle ndarray destination
-    if isinstance(out, np.ndarray):
+    # obtain ndarrays, casting to the same data type
+    buf = ensure_ndarray_from_memory(buf).view('u1')
+    out = ensure_ndarray_from_memory(out).view('u1')
 
-        # view source as destination dtype
-        if isinstance(buf, np.ndarray):
-            buf = buf.view(dtype=out.dtype).reshape(-1, order='A')
-        else:
-            buf = np.frombuffer(buf, dtype=out.dtype)
+    # copy memory
+    np.copyto(out, buf)
 
-        # ensure shapes are compatible
-        if buf.shape != out.shape:
-            if out.flags.f_contiguous:
-                order = 'F'
-            else:
-                order = 'C'
-            buf = buf.reshape(out.shape, order=order)
+    return out
 
-        # copy via numpy
-        np.copyto(out, buf)
 
-    # handle generic buffer destination
-    else:
+def ensure_text(l, encoding='utf-8'):
+    if isinstance(l, text_type):
+        return l
+    else:  # pragma: py3 no cover
+        return text_type(l, encoding=encoding)
 
-        # obtain memoryview of destination
-        dest = memoryview(out)
 
-        # ensure source is 1D
-        if isinstance(buf, np.ndarray):
-            buf = buf.reshape(-1, order='A')
-            # try to match itemsize
-            dtype = 'u%s' % dest.itemsize
-            buf = buf.view(dtype=dtype)
+def ensure_ndarray_from_memory(o, flatten=True):
+    """Convenience function to obtain a numpy ndarray using memory exposed by object `o`,
+    ensuring that no memory copies are made, and that `o` is not an object array.
 
-        # try to copy via memoryview
-        dest[:] = buf
+    Parameters
+    ----------
+    o : bytes-like
+        Any object exposing a memory buffer. On Python 3 this must be an object exposing
+        the new-style buffer interface. On Python 2 this can also be an object exposing
+        the old-style buffer interface.
+    flatten : bool, optional
+        If True, flatten any multi-dimensional inputs into a one-dimensional memoryview.
 
-    return out
+    Returns
+    -------
+    o : ndarray
 
+    """
 
-def ndarray_from_buffer(buf, dtype):
-    if isinstance(buf, np.ndarray):
-        arr = buf.reshape(-1, order='A').view(dtype)
-    else:
-        arr = np.frombuffer(buf, dtype=dtype)
-    return arr
+    if not isinstance(o, np.ndarray):
 
+        # first try to obtain a memoryview or buffer, needed to ensure that we don't
+        # subsequently copy memory when going via np.array()
 
-def ensure_text(l, encoding='utf-8'):
-    if isinstance(l, text_type):
-        return l
-    else:  # pragma: py3 no cover
-        return text_type(l, encoding=encoding)
+        if PY2:  # pragma: py3 no cover
```

Comment:
What's the advantage of the `if` over just the `try...catch`?

Reply:
FWIW I'm trying to create a clean separation for all PY2 code paths. The try...catch is only needed under PY2. So it will be obvious this whole conditional block can be deleted when PY2 is dropped.

Reply:
Yeah I can see how there is a little confusion either way. If this is preferred, let's go with it.